src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -1,197 +1,599 @@
+                // Narrow down to the smallest range [0, 2^bits] that contains maxValue.
+                // Then repeatedly generate a value in that outer range until we get one within the inner range.
+                int bits = BitOperations.Log2((ulong)maxValue) + 1;
+                while (true)
+                {
+                    ulong result = NextUInt64() >> (sizeof(long) * 8 - bits);
+                    if (result < (ulong)maxValue)
+                    {
+                        return (long)result;
+                    }
+                }
             }
-            if (retVal < 0)
+
+            public override long NextInt64(long minValue, long maxValue)
             {
-                retVal += int.MaxValue;
+                // Narrow down to the smallest range [0, 2^bits] that contains maxValue.
+                // Then repeatedly generate a value in that outer range until we get one within the inner range.
+                ulong exclusiveRange = (ulong)(maxValue - minValue);
+                int bits = BitOperations.Log2(exclusiveRange) + 1;
+                while (true)
+                {
+                    ulong result = NextUInt64() >> (sizeof(long) * 8 - bits);
+                    if (result < exclusiveRange)
+                    {
+                        return (long)result + minValue;
+                    }
+                }
             }
 
-            _seedArray[locINext] = retVal;
-            _inext = locINext;
-            _inextp = locINextp;
+            public override void NextBytes(byte[] buffer) => NextBytes((Span<byte>)buffer);
 
-            return retVal;
+            public override void NextBytes(Span<byte> buffer)
+            {
+                ulong next;
+
+                while (buffer.Length >= sizeof(long))
+                {
+                    next = NextUInt64();
+                    MemoryMarshal.Write(buffer, ref next);
+                    buffer = buffer.Slice(sizeof(long));
+                }
+
+                next = NextUInt64();
+                unsafe
+                {
+                    byte* remainingBytes = (byte*)&next;
+                    for (int i = 0; i < buffer.Length; i++)
+                    {
+                        buffer[i] = remainingBytes[i];
+                    }
+                }
+            }
+
+            public override double NextDouble() =>
+                // As described in http://prng.di.unimi.it/:
+                // "A standard double (64-bit) floating-point number in IEEE floating point format has 52 bits of significand,
+                //  plus an implicit bit at the left of the significand. Thus, the representation can actually store numbers with
+                //  53 significant binary digits. Because of this fact, in C99 a 64-bit unsigned integer x should be converted to
+                //  a 64-bit double using the expression
+                //  (x >> 11) * 0x1.0p-53"
+                (NextUInt64() >> 11) * (1.0 / (1ul << 53));
```

Comment:
Just thinking aloud here...

This will return a value between `0` and `0.99999999999999988897769753748434595763683319091796875`.

I'm pretty sure there will be no conflicts due to how `float`/`double` are represented (and validated on `float`).
However, for `float`, this is only `16m` unique values that will be exposed for while there are actually `1,065,353,216` (just over 1 billion) "unique" values between `0.0f` and `1.0f` (`1,056,964,608` if you exclude subnormal values).

The distribution of these 16m unique values is:
![Picture1](https://user-images.githubusercontent.com/10487869/104832386-8dcde200-5845-11eb-876d-13760a899cf4.png)

If we went across the entire `0.0f` to `1.0f` range, we end up with a distribution that looks more like:
![Picture2](https://user-images.githubusercontent.com/10487869/104832422-ce2d6000-5845-11eb-91d7-b5dc8ab006d6.png)

Likewise, for the entire `0.0f` to `+Infinity` range:
![Picture3](https://user-images.githubusercontent.com/10487869/104832428-d71e3180-5845-11eb-8170-0adf510d5320.png)

Reply:
So it's likely good that we are limiting ourselves to only 16m evenly distributed values and also that we aren't depending on `NextDouble()` or `NextSingle()` for distributing across a range of inputs, particular with the introduction of 64-bits.

For `double` this is be `2^53` evenly distributed values and we would have hit other limits with representable integers over `2^52` (which likely would have caused issues for `NextInt64(range)`.

Reply:
See NextDoubleHighRes() on https://github.com/colgreen/Redzen/blob/master/Redzen/Random/RandomSourceBase.cs for an algorithm that samples in the interval [0,1) using all of the possible values in that interval provided by a IEEE754 double. 

As you will see, that method is rather more complex than the 'standard' method of just sampling from the uniformly distributed subset of possible values, and as such is very much slower. I added NextDoubleHighRes() as a sort of code golf exercise to be honest, as I suspect the use cases for it are rare and specialized.